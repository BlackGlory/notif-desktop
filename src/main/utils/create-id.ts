import { timestampBasedId } from 'extra-generator'

const iter = timestampBasedId()

export function createTimeBasedId(): [timestamp: number, num: number] {
  return iter.next().value
}

export function stringifyTimeBasedId(
  [timestamp, num]: [timestamp: number, num: number]
): string {
  const timestampString = timestamp.toString().padStart(20, '0')
  const numString = num.toString()
  return `${timestampString}-${numString}`
}
