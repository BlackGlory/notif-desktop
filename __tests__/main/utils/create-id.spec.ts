import { describe, test, expect } from 'vitest'
import { stringifyTimeBasedId } from '@main/utils/create-id.js'
import { assert } from '@blackglory/prelude'

describe('stringifyTimeBasedId([timestamp: number, num: number]): string', () => {
  test('compare', () => {
    assert(stringifyTimeBasedId([0, 0]) < stringifyTimeBasedId([0, 1]), 'expect (0,0) < (0,1)')
    assert(stringifyTimeBasedId([0, 1]) < stringifyTimeBasedId([0, 10]), 'expect (0,1) < (0,10)')
    assert(stringifyTimeBasedId([0, 10]) < stringifyTimeBasedId([0, 999]), 'expect (0,10) < (0,999)')
    assert(stringifyTimeBasedId([0, 999]) < stringifyTimeBasedId([1, 0]), 'expect (0,999) < (1,0)')

    assert(stringifyTimeBasedId([1, 0]) < stringifyTimeBasedId([1, 1]), 'expect (1,0) < (1,1)')
    assert(stringifyTimeBasedId([1, 1]) < stringifyTimeBasedId([1, 10]), 'expect (1,1) < (1,10)')
    assert(stringifyTimeBasedId([1, 10]) < stringifyTimeBasedId([1, 999]), 'expect (1,10) < (1,999)')
    assert(stringifyTimeBasedId([1, 999]) < stringifyTimeBasedId([2, 0]), 'expect (1,999) < (2,0)')

    assert(stringifyTimeBasedId([1, 0]) < stringifyTimeBasedId([10, 0]), 'expect (1,0) < (10,0)')
    assert(stringifyTimeBasedId([1, 1]) < stringifyTimeBasedId([10, 0]), 'expect (1,1) < (10,0)')
    assert(stringifyTimeBasedId([1, 10]) < stringifyTimeBasedId([10, 0]), 'expect (1,10) < (10,0)')
    assert(stringifyTimeBasedId([1, 999]) < stringifyTimeBasedId([10, 0]), 'expect (1,999) < (10,0)')

    assert(stringifyTimeBasedId([9, 0]) < stringifyTimeBasedId([10, 0]), 'expect (9,0) < (10,0)')
    assert(stringifyTimeBasedId([9, 1]) < stringifyTimeBasedId([10, 0]), 'expect (9,1) < (10,0)')
    assert(stringifyTimeBasedId([9, 10]) < stringifyTimeBasedId([10, 0]), 'expect (9,10) < (10,0)')
    assert(stringifyTimeBasedId([9, 999]) < stringifyTimeBasedId([10, 0]), 'expect (9,999) < (10,0)')

    assert(stringifyTimeBasedId([10, 0]) < stringifyTimeBasedId([10, 1]), 'expect (10,0) < (10,1)')
    assert(stringifyTimeBasedId([10, 1]) < stringifyTimeBasedId([10, 999]), 'expect (10,1) < (10,999)')
    assert(stringifyTimeBasedId([10, 999]) < stringifyTimeBasedId([11, 0]), 'expect (10,999) < (11,0)')
  })

  test('snapshot', () => {
    expect(stringifyTimeBasedId([0, 0])).toMatchSnapshot()
    expect(stringifyTimeBasedId([0, 1])).toMatchSnapshot()
    expect(stringifyTimeBasedId([0, 10])).toMatchSnapshot()
    expect(stringifyTimeBasedId([0, 999])).toMatchSnapshot()

    expect(stringifyTimeBasedId([1, 0])).toMatchSnapshot()
    expect(stringifyTimeBasedId([1, 1])).toMatchSnapshot()
    expect(stringifyTimeBasedId([1, 10])).toMatchSnapshot()
    expect(stringifyTimeBasedId([1, 999])).toMatchSnapshot()

    expect(stringifyTimeBasedId([9, 0])).toMatchSnapshot()
    expect(stringifyTimeBasedId([9, 1])).toMatchSnapshot()
    expect(stringifyTimeBasedId([9, 10])).toMatchSnapshot()
    expect(stringifyTimeBasedId([9, 999])).toMatchSnapshot()

    expect(stringifyTimeBasedId([10, 0])).toMatchSnapshot()
    expect(stringifyTimeBasedId([10, 1])).toMatchSnapshot()
    expect(stringifyTimeBasedId([10, 999])).toMatchSnapshot()
  })
})
