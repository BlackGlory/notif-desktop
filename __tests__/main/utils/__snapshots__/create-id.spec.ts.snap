// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 1`] = `"00000000000000000000-0"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 2`] = `"00000000000000000000-1"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 3`] = `"00000000000000000000-10"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 4`] = `"00000000000000000000-999"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 5`] = `"00000000000000000001-0"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 6`] = `"00000000000000000001-1"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 7`] = `"00000000000000000001-10"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 8`] = `"00000000000000000001-999"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 9`] = `"00000000000000000009-0"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 10`] = `"00000000000000000009-1"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 11`] = `"00000000000000000009-10"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 12`] = `"00000000000000000009-999"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 13`] = `"00000000000000000010-0"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 14`] = `"00000000000000000010-1"`;

exports[`stringifyTimeBasedId([timestamp: number, num: number]): string > snapshot 15`] = `"00000000000000000010-999"`;
